### ISSUE TITLE
Direct build of layer/attention silently produces non-causal output

### ISSUE THREAD

**Issue by ayghri** [huggingface/transformers#46273]
### Feature request

- Option A: make the eager fallback honor `module.is_causal`:

```python
def eager_attention_forward(module, query, key, value, attention_mask, scaling, dropout=0.0, **kwargs):
    ...
    attn_weights = torch.matmul(query, key_states.transpose(2, 3)) * scaling
    if attention_mask is not None:
        attn_weights = attn_weights + attention_mask
    elif getattr(module, "is_causal", False):           # apply causal mask when none is supplied
        L, S = query.shape[-2], key_states.shape[-2]
        causal = torch.ones(L, S, dtype=torch.bool, device=query.device).tril(diagonal=S - L)
        attn_weights = attn_weights.masked_fill(~causal, float("-inf"))
    ...
```

- Option B: let users export the fully-resolved config: doesn't change attention behavior, but gives a way to snapshot and restore the resolved backend.

```python
config.to_dict(include_runtime=True)         # keep _attn_implementation, etc.
config.to_json_string(include_runtime=True)
```
```python
def _remove_keys_not_serialized(self, d, include_runtime: bool = False):
    keys_to_remove = ["_is_quantized", "_auto_class", "_commit_hash", ...]
    if not include_runtime:
        keys_to_remove += ["_attn_implementation_internal", "_experts_implementation_internal"]
    for key_to_remove in keys_to_remove:
        d.pop(key_to_remove, None)
    ...
```

These are not mutually exclusive, I think B improves reproducibility. Happy to open a PR for whichever direction maintainers prefer.

### Motivation

When a decoder module is built directly from a config instead of via `AutoModel.from_pretrained` (using this for distillation), `config._attn_implementation` is `None`:

```python
cfg = AutoConfig.from_pretrained("Qwen/Qwen3-4B")
cfg._attn_implementation            # -> None  (never set without a model)
layer = Qwen3DecoderLayer(cfg, 0)   # dispatch falls back to eager
```

`AttentionInterface.get_interface(None, ...)` falls back to `eager_attention_forward`, which only masks `if attention_mask is not None`. With `attention_mask=None`, the standalone module computes **bidirectional** attention instead of causal.

The module (for Qwen3) already declares itself causal (`self.is_causal = True`), and `to_dict()` / `to_json_string()` strip the resolved backend unconditionally (`_remove_keys_not_serialized`), so there is no way to round-trip or recover it either.

### Minimal reproduction

```python
import torch
from transformers import Qwen3Config
from transformers.models.qwen3.modeling_qwen3 import Qwen3Attention, Qwen3RotaryEmbedding

cfg = Qwen3Config(hidden_size=64, num_attention_heads=4, num_key_value_heads=2,
                  head_dim=16, intermediate_size=128, num_hidden_layers=1, vocab_size=100)

assert cfg._attn_implementation is None 
attn = Qwen3Attention(cfg, layer_idx=0).eval()
assert attn.is_causal is True   

B, T = 1, 8
x = torch.randn(B, T, cfg.hidden_size)
pe = Qwen3RotaryEmbedding(cfg)(x, torch.arange(T).unsqueeze(0))

with torch.no_grad():
    out_none, _ = attn(x, position_embeddings=pe, attention_mask=None)  # None -> eager, NOT causal
    cfg._attn_implementation = "sdpa"
    out_sdpa, _ = attn(x, position_embeddings=pe, attention_mask=None)  # sdpa, is_causal=True

print((out_none - out_sdpa).norm())   # large: bidirectional vs causal
```

### Your contribution

I can submit PR for either Option A or B. I'm currently relying on B in my local monkey patch.

**Comment by anshuS1310** [comment]
I think option A is good because it corrects the eager fallback mechanism. would option A be better ?

**Comment by zucchini-nlp** [comment]
Interesting issue. I suppose it works well and returns the same output when init a whole text model, because we usually prepare a mask based on model's config and inputs. So the issue is when an attention module is called in isolatation 🤔 

TBH I also think the first option makes more sense, though it will be a huge and breaking change, touching all models. Tagging @vasqu for it

**Comment by anshuS1310** [comment]
Can I  PR for this ? @zucchini-nlp

**Comment by zucchini-nlp** [comment]
Lets wait for @vasqu , even though the option A sounds reasonable, it will be a huge breaking change so I don't want us to commit, before making sure it is absolutely the right direction

**Comment by anshuS1310** [comment]
ohk l

**Comment by vasqu** [comment]
Ok sorry but I have to backtrack on these options.

These flags are not saved for a reason as they are dynamically resolved during `from_pretrained` (checking your env and passed kwargs). Having only a decoder layer makes this impossible because now we are outside the realm of `PreTrainedModel` (there is a reason they are not exposed in the base `__init__`). Ofc, you can statically force this to be included (B) but this is in no way safe and relies on the user what to do, which I'm not a fan of. Option A is also not really viable because this is an explicit design choice to rely on the mask to be properly created before (this is very similar to SDPA); otherwise you just think of the easiest case and already ignore SWA.

Your use case just seems very special tbh. Imo, the best way in your case is to create the mask yourself, the mask API is simple enough tbh and force the config to have some attention flavor - you are sure you have available (because we can no longer track this outside of a pretrained model). Another option would be to properly use the pretrained module (with an adjusted config) and monkey-patch what you dont need with `nn.Identity`

**Comment by anshuS1310** [comment]
is this open ? can I work on this ? 
@vasqu

**Comment by ayghri** [comment]
Sounds fair. Although I still find Option B more viable: it's 100% backward-compatible, and helps when the config is saved from same env/hardware as the training node. But I get the logic behind maintaining current behavior.

@vasqu @anshuS1310 @zucchini-nlp 

I'm closing this for now.

**Comment by anshuS1310** [comment]
Ok🥲🥲🥲
